concordium-consensus/src/Concordium/KonsensusV1/Consensus.hs

```diff
@@ -325,3 +327,52 @@ getProtocolUpdateState = do
                     { puQueuedTime = ts,
                       puProtocolUpdate = pu
                     }
+
+-- |Project the earliest timestamp at which the given baker might be required to bake.
+-- If the baker is not a baker in the current reward period, this will give a time at the start
+-- of the next reward period.
+bakerEarliestWinTimestamp ::
+    ( GSTypes.BlockState m ~ PBS.HashedPersistentBlockState (MPV m),
+      BS.BlockStateQuery m,
+      IsConsensusV1 (MPV m)
+    ) =>
+    BakerId ->
+    SkovData (MPV m) ->
+    m Timestamp
+bakerEarliestWinTimestamp baker sd = do
+    let lfBlock = sd ^. lastFinalized
+    -- The bakers with respect to the epoch of the last finalized block.
+    let fullBakers = sd ^. currentEpochBakers . bfBakers
+    let epochDuration = genesisEpochDuration . gmParameters $ sd ^. genesisMetadata
+    lfSeedState <- BS.getSeedState (bpState lfBlock)
+    let lfTriggerTime = lfSeedState ^. triggerBlockTime
+    chainParams <- BS.getChainParameters (bpState lfBlock)
+    let minBlockTime = chainParams ^. cpConsensusParameters . cpMinBlockTime
+    if null (fullBaker fullBakers baker)
+        then do
+            -- The baker is not a baker in the current payday, so take the start of the next payday
+            -- as the soonest that the account could be a baker (and thus bake).
+            let epochsTillPayday = (sd ^. nextPayday) - blockEpoch lfBlock - 1
+            return $!
+                addDuration
+                    lfTriggerTime
+                    (minBlockTime + fromIntegral epochsTillPayday * epochDuration)
+        else do
+            let findLeader nxtRound nxtTimestamp
+                    | baker
+                        == getLeaderFullBakers
+                            fullBakers
+                            (lfSeedState ^. currentLeadershipElectionNonce)
+                            nxtRound
+                            ^. Accounts.bakerIdentity =
+                        nxtTimestamp
+                    | nxtTimestamp >= lfTriggerTime = addDuration lfTriggerTime minBlockTime
+                    | otherwise = findLeader (nxtRound + 1) (addDuration nxtTimestamp minBlockTime)
```

Comment:
Regarding the performance - would it make sense to introduce a configurable "max lookahead" ? On the other hand it would probably just make things needlessly complex for the consumer...

Reply:
I would prefer not to if it performs good enough. It just complicates the API.

Reply:
I've added a bypass for when the minimum block time is 0. On my machine, the query takes about 1 second if it has to look around 3,000,000 rounds ahead. With an epoch length of 1 hour and a minimum block time of 1 second, it only has to look 3,600 rounds ahead at most. So in practice, this should be safe. (On my machine, running the query in concordium-client took up to about 400 ms with that kind of look ahead.)

Reply:
I'm a bit surprised that looking 3000000 rounds ahead takes 1second, but the query still takes 400ms.

I wonder where the overhead is then.

Reply:
I think it's because I was running it was `stack exec`, which adds a lot of overhead.